pkg/codegen/go/gen.go

```diff
@@ -2513,6 +2563,73 @@ func generatePackageContextMap(tool string, pkg *schema.Package, goInfo GoPackag
 		scanResource(r)
 	}
 
+	// compute set of names generated by a type
+	// handling any potential collisions via remapping along the way
+	scanType := func(t schema.Type) {
+		switch t := t.(type) {
+		case *schema.ObjectType:
+			pkg := getPkgFromToken(t.Token)
+			// maintain support for duplicate tokens for types and resources in Kubernetes
+			if resSeen[t.Token] {
+				pkg.duplicateTokens[t.Token] = true
+				return
+			}
+			if typeSeen[t.Token] {
+				return
+			}
+			typeSeen[t.Token] = true
+
+			name := pkg.tokenToType(t.Token)
+			getNames := func(suffix string) []string {
+				names := []string{}
+				names = append(names, name+suffix)
+				names = append(names, name+suffix+"Input")
+				names = append(names, name+suffix+"Output")
+				return names
+			}
+			suffixes := []string{"", "Type", "Typ"}
+			suffix := ""
+			suffixIndex := 0
+			canGenerate := false
+
+			for !canGenerate && suffixIndex <= len(suffixes) {
+				suffix = suffixes[suffixIndex]
+				candidates := getNames(suffix)
+				conflict := false
+				for _, c := range candidates {
+					if pkg.names.Has(c) {
+						conflict = true
+					}
+				}
+				if !conflict {
+					canGenerate = true
+					break
+				}
+
+				suffixIndex++
+			}
+
+			if !canGenerate {
+				panic(fmt.Sprintf("unable to generate Go SDK, schema has unresolvable overlapping type: %s", name))
+			}
+
+			names := getNames(suffix)
+			originalNames := getNames("")
+			for i, n := range names {
+				pkg.names.Add(n)
+				if suffix != "" {
+					pkg.renamed[originalNames[i]] = names[i]
+				}
+			}
+		default:
+			return
+		}
+	}
+
+	for _, t := range pkg.Types {
```

Comment:
Is this deterministic or we need a sort here --^

Reply:
The schema binder handles sorting resource and type arrays before we get to this point: https://github.com/pulumi/pulumi/blob/76ee1b8ccfee815eb315d9e0e0ddfaaf505c472b/pkg/codegen/schema/schema.go#L1506